USB DFU fixes

aurel32 opened the pull request:
This patchset fixes the DFU firmware upload function and the support for DFU in composite mode. It also fixes the upload code in view of unifying the USB legacy and composite code (issue #10503).

Fixes #12813

aurel32 committed:
usb: class: dfu: fix firmware upload

The flash_read() function takes an offset from beginning of the flash.
This patch subtract FLASH_BASE_ADDRESS from the absolute address. This
fixes DFU firmware upload.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: class: dfu: fix USB descriptors in composite mode

In composite mode the request handler buffer is common to all functions,
and its size is defined by CONFIG_USB_COMPOSITE_BUFFER_SIZE.

Given the device is enumerated as composite in runtime mode and
as legacy in download mode, we need to define wTransferSize as
the minimum of the two.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: class: dfu: use the request handler provided buffer

The upload code assumes the buffer to use is the DFU provide one, and
not the one provided by the request handler, which is different in
composite mode.

This is only a theoretical issue, as this function is only executed
once the device has been re-enumerated in legacy mode, but that is a
required step before unifying legacy and composite code.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: class: dfu: always provide payload_data

Given the device can be enumerated as both legacy and composite, we
always need to provide the payload_data buffer.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

nashif commented:
no need for "subsys: " prefix in commits :)

aurel32 commented:
> no need for "subsys: " prefix in commits :)

Ok, I just pushed a new version without it.

aurel32 commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -480,7 +480,8 @@ static int dfu_class_handle_req(struct usb_setup_packet *pSetup,
 			if (len) {
 				ret = flash_read(dfu_data.flash_dev,
-						 dfu_data.flash_addr +
+						 dfu_data.flash_addr -
```
> I was just confused why you want to do it, and I looked at the flash drivers for stm32, gecko and sam. Whatever this flash drivers do, I do not agree with the change here.

For what I understand, flash drivers take an offset relative to the beginning of the flash and not an absolute address. This is the case of at least the STM32 driver and the Kinetis MCU driver:
https://github.com/zephyrproject-rtos/zephyr/blob/7a85c19f1a01c56a0f0b2fed47b464471e348d6d/drivers/flash/soc_flash_mcux.c#L70

The NRF driver doesn't add any offset, but on the other hand the flash is mapped at address 0.

When writing data to the flash through `flash_img_buffered_write()` which in turns call `flash_write()` an offset relative to the beginning of the flash (`FLASH_AREA_IMAGE_1_OFFSET`) is used. On the other hand when the DFU code calls `flash_read`, it passes an absolute address as `dfu_data.flash_addr` is assigned either as `CONFIG_FLASH_BASE_ADDRESS +  FLASH_AREA_IMAGE_0_OFFSET`or  `CONFIG_FLASH_BASE_ADDRESS +  FLASH_AREA_IMAGE_1_OFFSET`.This is not consistent.

> It will at least break DFU on Kinetis MCUs (and NRF?, arduino101?).

Unfortunately I don't have a Kinetis board to try. Can you please tell me what would break on Kinetis MCUs, especially given CONFIG_FLASH_BASE_ADDRESS is set to 0 on those boards?

aurel32 replied:
@jfischer-phytec-iot any hint why it will break DFU on Kinetis MCUs?

jfischer-phytec-iot replied:
@aurel32 That's not as dramatic as I thought, I did not remember where FLASH_BASE_ADDRESS came from. As I see all the drivers where the reg for the flash node is zero just left it out, it not consistent. I am still confuse why it is not already DT_FLASH_BASE_ADDRESS, @galak ?

galak replied:
> We would need the help from @galak to understand the DT part.

What's the question?  Is this about what the flash drivers are doing or something else?  We should be using CONFIG_FLASH_BASE_ADDRESS (that's what we generate from DTS).

aurel32 replied:
@galak Thanks for the confirmation, that's what is used in the patch.

jfischer-phytec-iot replied:
please consider discussion in  #12813

carlescufi commented:
@aurel32 @jfischer-phytec-iot what remains to be done here?

bfrog commented:
Would love to have this working in 1.14

aurel32 commented:
> @aurel32 @jfischer-phytec-iot what remains to be done here?
We would need the help from @galak to understand the DT part.

galak commented:
> We would need the help from @galak to understand the DT part.

what do you need help to understand?  Ping me on slack if that's easier.

carlescufi commented:
@aurel32 do you have plans to fix this for 1.14?

aurel32 commented:
> @aurel32 do you have plans to fix this for 1.14?

I think there is nothing to fix, it just needs a rebase. I have done one locally, but I still need to test it.

aurel32 commented:
@jfischer-phytec-iot I have just rebased the PR.

jfischer-phytec-iot commented:
@aurel32 would you apply @nashif suggestion from #12813?

aurel32 commented:
> @aurel32 would you apply @nashif suggestion from #12813?

I am just discovering that issue now. I'll add that patch in the PR.

aurel32 committed:
usb: class: dfu: fix support for SPI based flash

With an SPI based flash, CONFIG_FLASH_BASE_ADDRESS is not defined. In
that case it is safe to assume that the base address is 0.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
> > @aurel32 would you apply @nashif suggestion from #12813?
> 
> I am just discovering that issue now. I'll add that patch in the PR.

Done.
